ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3HeadHandler.java

```diff
@@ -0,0 +1,180 @@
+      bucketHandler.handle(restRequest, restResponseChannel, wrappedCallback);
+    } else {
+      objectHandler.handle(restRequest, restResponseChannel, wrappedCallback);
+    }
+  }
+
+  private class S3HeadBucketHandler {
+    private final Logger LOGGER = LoggerFactory.getLogger(S3HeadBucketHandler.class);
+    private final SecurityService securityService;
+    private final FrontendMetrics metrics;
+    private final AccountService accountService;
+
+    private S3HeadBucketHandler(SecurityService securityService, FrontendMetrics metrics,
+        AccountService accountService) {
+      this.securityService = securityService;
+      this.metrics = metrics;
+      this.accountService = accountService;
+    }
+
+    private void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<ReadableStreamChannel> callback) {
+      if (!getRequestPath(restRequest).matchesOperation(Operations.NAMED_BLOB)) {
+        throw new RuntimeException("S3HeadBucketHandler should only handle named blob requests");
+      }
+      new CallbackChain(restRequest, restResponseChannel, callback).start();
+    }
+
+    private class CallbackChain {
+      private final RestRequest restRequest;
+      private final RestResponseChannel restResponseChannel;
+      private final Callback<ReadableStreamChannel> finalCallback;
+
+      private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+          Callback<ReadableStreamChannel> finalCallback) {
+        this.restRequest = restRequest;
+        this.restResponseChannel = restResponseChannel;
+        this.finalCallback = finalCallback;
+      }
+
+      private void start() {
+        restRequest.setArg(RestUtils.InternalKeys.KEEP_ALIVE_ON_ERROR_HINT, true);
+        securityService.processRequest(restRequest, securityProcessRequestCallback());
+      }
+
+      private Callback<Void> securityProcessRequestCallback() {
+        return buildCallback(metrics.headBlobSecurityProcessRequestMetrics,
+            result -> securityService.postProcessRequest(restRequest, securityPostProcessRequestCallback()),
+            restRequest.getUri(), LOGGER, null);
+      }
+
+      private Callback<Void> securityPostProcessRequestCallback() {
+        return buildCallback(metrics.getAccountsSecurityPostProcessRequestMetrics, securityCheckResult -> {
+          RequestPath requestPath = (RequestPath) restRequest.getArgs().get(REQUEST_PATH);
+          // Build a dummy blob path to get the account and container name
+          String dummyBlobPath = requestPath.getOperationOrBlobId(true) + "/dummy";
+          NamedBlobPath namedBlobPath = NamedBlobPath.parse(dummyBlobPath, restRequest.getArgs());
+          String accountName = namedBlobPath.getAccountName();
+          String containerName = namedBlobPath.getContainerName();
+          try {
+            Container container = accountService.getContainerByName(accountName, containerName);
```

Comment:
For the container, if we go with the fixed container s3_container_a or so?
Is this a virtual container or physical container which exists in the database?

Reply:
I think this would need to be a physical container. Later, I think we could have different containers for different level of services such as encryption, compaction, etc. @Arun-LinkedIn could you comment?

Reply:
@Arun-LinkedIn Ack.